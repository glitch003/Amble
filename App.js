import './shim.js'
import React, { Component } from 'react'
import {
  Platform,
  StyleSheet,
  Text,
  View,
  TouchableOpacity
} from 'react-native'

import { DrawerNavigator, StackNavigator } from 'react-navigation'
import Ionicons from 'react-native-vector-icons/Ionicons'

import Home from './scenes/Home'
import Notifications from './scenes/Notifications'
import PicEncrypt from './scenes/PicEncrypt'

// use this to debug the JS bridge
// require('MessageQueue').spy(true)

const headerLeft = (navigation) => {
  return (
    <TouchableOpacity
      onPress={() => navigation.navigate('DrawerToggle')}>
      <Ionicons
        name='md-menu'
        size={20}
        style={{paddingLeft: 10}}
      />
    </TouchableOpacity>
  )
}

const homeScreenStack = StackNavigator(
  {
    Home: {
      screen: Home
    }
  },
  {
    navigationOptions: ({ navigation }) => ({
      initialRouteName: 'Home',
      headerMode: 'screen',
      headerTitle: 'Amble',
      drawerLabel: 'Amble',
      drawerIcon: ({ tintColor, focused }) => (
        <Ionicons
          name={focused ? 'ios-home' : 'ios-home-outline'}
          size={20}
          style={{ color: tintColor }}
        />
      ),
      headerLeft: headerLeft(navigation)
    })
  }
)

const notificationScreenStack = StackNavigator(
  {
    Notifications: {
      screen: Notifications
    }
  },
  {
    navigationOptions: ({ navigation }) => ({
      initialRouteName: 'Notifications',
      headerMode: 'screen',
      headerTitle: 'Notifications',
      drawerLabel: 'Notifications',
      drawerIcon: ({ tintColor, focused }) => (
        <Ionicons
          name={focused ? 'ios-notifications' : 'ios-notifications-outline'}
          size={20}
          style={{ color: tintColor }}
        />
      ),
      headerLeft: headerLeft(navigation)
    })
  }
)

const picEncryptStack = StackNavigator(
  {
    PicEncrypt: {
      screen: PicEncrypt
    }
  },
  {
    navigationOptions: ({ navigation }) => ({
      initialRouteName: 'PicEncrypt',
      headerMode: 'screen',
      headerTitle: 'Encrypt Photo',
      drawerLabel: 'Encrypt Photo',
      drawerIcon: ({ tintColor, focused }) => (
        <Ionicons
          name={focused ? 'ios-camera' : 'ios-camera-outline'}
          size={20}
          style={{ color: tintColor }}
        />
      ),
      headerLeft: headerLeft(navigation)
    })
  }
)

const App = DrawerNavigator(
  {
    Home: {
      screen: homeScreenStack
    },
    Notifications: {
      screen: notificationScreenStack
    },
    PicEncrypt: {
      screen: picEncryptStack
    }
  }
)

export default App
