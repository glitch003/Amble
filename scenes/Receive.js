import React, { Component } from 'react'
import {
  Platform,
  StyleSheet,
  Text,
  View,
  TouchableOpacity,
  Clipboard
} from 'react-native'
import theme from '../config/Theme'

export default class Receive extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      addressString: props.address || global.currentWallet.getAddressString()
    }
  }
  render () {
    return (
      <View style={styles.container}>
        <TouchableOpacity onPress={() => Clipboard.setString(this.state.addressString)}>
          <Text style={styles.yourAddress}>
            Your address:
          </Text>
          <Text
            style={styles.welcome}
            adjustsFontSizeToFit
            allowFontScaling
            minimumFontScale={1}
            numberOfLines={1}>
            {this.state.addressString}
          </Text>
        </TouchableOpacity>
        <View style={styles.qrCode}>
          {global.currentWallet.renderAddressQRCode(this.state.addressString)}
        </View>
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'flex-start',
    alignItems: 'center',
    backgroundColor: theme.sceneBackgroundColor
  },
  welcome: {
    textAlign: 'center'
  },
  yourAddress: {
    fontSize: 15,
    textAlign: 'center',
    marginTop: 20,
    marginBottom: 10
  },
  qrCode: {
    marginTop: 20
  }
})
