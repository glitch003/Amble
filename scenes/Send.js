import React from 'react'
import {
  StyleSheet,
  Text,
  View,
  TextInput,
  TouchableOpacity,
  Image,
  Linking,
  Alert
} from 'react-native'
import { NavigationActions } from 'react-navigation'
import theme from '../config/Theme'
import AmbleButton from '../components/AmbleButton'

export default class Send extends React.Component {
  constructor () {
    super()
    this.state = {
      to: '',
      value: '',
      data: '',
      buttonState: 'send'
    }
  }

  componentDidMount () {
    console.log('mounting send: ' + JSON.stringify(this.props.navigation.state.params))
    if (this.props.navigation.state.params && this.props.navigation.state.params.txData) {
      const txData = this.props.navigation.state.params.txData
      this.setState({
        to: txData.to,
        value: String(txData.value),
        data: txData.data
      })
    }
  }

  renderDataView () {
    return (
      <View style={styles.horizontalContainer}>
        <Text style={styles.formLabel}>
          Data:
        </Text>
        <TextInput
          style={{height: 40, width: 200, borderColor: 'gray', borderWidth: 1, backgroundColor: 'white'}}
          onChangeText={(data) => this.setState({data})}
          value={this.state.data}
        />
      </View>
    )
  }

  render () {
    // use this snippet as an example of how you can recover a user's account from a QR code that was sent to their email
    // if (this.state.recovering) {
    //   return this.state.wallet.renderRecoveryQRScanner(() => {
    //     console.log('wallet has been recovered')
    //     console.log('address is ' + this.state.wallet.getAddressString())
    //     this.setState({recovering: false})
    //     this.state.wallet.getBalance()
    //     .then(balance => this.setState({balance}))
    //   })
    // }
    return (
      <View style={styles.container}>
        <WithLabel label='To' style={{marginTop: 30}}>
          <TextInput
            style={[styles.textInput, {height: 40}]}
            onChangeText={(to) => this.setState({to})}
            value={this.state.to}
            autoCapitalize='none'
            autoCorrect={false}
            multiline
            underlineColorAndroid='transparent'
          />
        </WithLabel>

        <WithLabel label='Value (ETH)'>
          <TextInput
            style={styles.textInput}
            onChangeText={(value) => this.setState({value})}
            value={this.state.value}
            autoCapitalize='none'
            autoCorrect={false}
            keyboardType='decimal-pad'
            underlineColorAndroid='transparent'
          />
        </WithLabel>

        {this.state.data === '' ? null : this.renderDataView()}

        <AmbleButton
          buttonProps={{
            buttonState: this.state.buttonState,
            states: {
              send: {
                onPress: () => {
                  this.setState({ buttonState: 'sending' })
                  this.sendTransaction()
                },
                text: 'Send'
              },
              sending: {
                onPress: () => {
                  this.setState({ buttonState: 'send' })
                },
                spinner: true,
                text: 'Sending...'
              }
            }
          }}
        />
      </View>
    )
  }

  sendTransaction () {
    // convert eth to wei because that's what sendTx takes
    let weiVal = global.etherUnits.toWei(this.state.value, 'ether')
    global.currentWallet.sendTx(this.state.to, weiVal, this.state.data)
    .then((result) => {
      // tell the user the txn was sent and offer them an option to view on etherscan
      Alert.alert(
        'Success',
        'Your transaction was sent!',
        [
          {
            text: 'Ok',
            onPress: () => {
              // pop to home
              this.props.navigation.dispatch(NavigationActions.back())
            },
            style: 'cancel'
          },
          {
            text: 'View on Etherscan',
            onPress: () => {
              // take the user to etherscan with tx id of result
              let url = global.sdkdConfig.moduleConfig.wallet.etherscanHost + '/tx/' + result
              // pop to home after opening link
              Linking.openURL(url)
              .then(result => {
                this.props.navigation.dispatch(NavigationActions.back())
              })
              .catch(err => console.error('An error occurred', err))
            }
          }
        ]
      )
    })
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'flex-start',
    alignItems: 'center',
    backgroundColor: theme.sceneBackgroundColor
  },
  welcome: {
    textAlign: 'center'
  },
  qrCode: {
    marginTop: 20
  },
  labelContainer: {
    flexDirection: 'row',
    marginVertical: 10,
    marginHorizontal: 10,
    height: 40
  },
  label: {
    width: 80,
    alignItems: 'flex-end',
    marginRight: 10,
    paddingTop: 2,
    height: 40
  },
  textInput: {
    height: 26,
    borderWidth: 1,
    borderColor: 'gray',
    flex: 1,
    fontSize: 13,
    padding: 4,
    backgroundColor: 'white'
  }
})

class WithLabel extends React.Component {
  render () {
    return (
      <View style={[styles.labelContainer, this.props.style]}>
        <View style={styles.label}>
          <Text>{this.props.label}</Text>
        </View>
        {this.props.children}
      </View>
    )
  }
}
