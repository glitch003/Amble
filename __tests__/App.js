import 'react-native'
import React from 'react'
import MockAsyncStorage from 'mock-async-storage'
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer'

import App from '../App'
import Send from '../scenes/Send'
import Receive from '../scenes/Receive'

jest.mock('WebView', () => 'WebView')

jest.mock('NativeModules', () => {
  return {
    RNRandomBytes: {
      seed: undefined,
      randomBytes: jest.fn()
    }
  }
})

jest.mock('react-native-keychain', () => {
  let domains = {}
  return {
    setInternetCredentials: jest.fn((domain, email, password) => {
      domains[domain] = {
        email: email,
        password: password
      }
      return Promise.resolve()
    }),
    getInternetCredentials: jest.fn((domain) => {
      return new Promise((resolve, reject) => {
        if (domains[domain] === null || domains[domain] === undefined || domains[domain].email == null) {
          resolve(false)
        }
        resolve({username: domains[domain].email, password: domains[domain].password})
      })
    })
  }
})

jest.mock('react-native-camera', () => {
  return {}
})

jest.mock('react-native-push-notification', () => {
  return {
    configure: jest.fn()
  }
})

const mockStorage = () => {
  const mockImpl = new MockAsyncStorage()
  jest.mock('AsyncStorage', () => mockImpl)
}

mockStorage()

it('whole app renders correctly', () => {
  const tree = renderer.create(
    <App />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

it('send screen renders correctly', () => {
  const tree = renderer.create(
    <Send
      navigation={{
        state: {
          params: {}
        }
      }}
    />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

it('receive screen renders correctly', () => {
  const tree = renderer.create(
    <Receive address='12345' />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})
