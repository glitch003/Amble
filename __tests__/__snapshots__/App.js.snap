// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`receive screen renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "justifyContent": "flex-start",
    }
  }
>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    collapsable={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 15,
          "marginBottom": 10,
          "marginTop": 20,
          "textAlign": "center",
        }
      }
    >
      Your address:
    </Text>
    <Text
      accessible={true}
      adjustsFontSizeToFit={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      minimumFontScale={1}
      numberOfLines={1}
      style={
        Object {
          "textAlign": "center",
        }
      }
    >
      12345
    </Text>
  </View>
  <View
    style={
      Object {
        "marginTop": 20,
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "height": 200,
            "width": 200,
          }
        }
      >
        <WebView
          automaticallyAdjustContentInsets={false}
          contentInset={
            Object {
              "bottom": 0,
              "left": 0,
              "right": 0,
              "top": 0,
            }
          }
          javaScriptEnabled={true}
          opaque={false}
          scalesPageToFit={false}
          scrollEnabled={false}
          source={
            Object {
              "html": "<style>*{margin:0;padding:0;}canvas{transform:translateZ(0);}</style><canvas></canvas><script>var canvas = document.querySelector('canvas');(function renderCanvas(canvas){
var ctx=canvas.getContext('2d');
var size=this.size;
var fgColor=this.fgColor;
var bgColor=this.bgColor;
canvas.width=size;
canvas.height=size;
canvas.style.left=(window.innerWidth-size)/2+'px';
if(window.innerHeight>size)canvas.style.top=(window.innerHeight-size)/2+'px';
ctx.fillRect(0,0,size,size);
var cells=this.cells;
var cellWidth=this.size/cells.length;
var cellHeight=this.size/cells.length;
var nRoundedWidth=Math.round(cellWidth);
var nRoundedHeight=Math.round(cellHeight);
cells.forEach(function(row,rowIndex){
row.forEach(function(column,columnIndex){
var nLeft=columnIndex*cellWidth;
var nTop=rowIndex*cellHeight;
ctx.fillStyle=ctx.strokeStyle=column?bgColor:fgColor;
ctx.lineWidth=1;
ctx.fillRect(nLeft,nTop,cellWidth,cellHeight);
ctx.strokeRect(
Math.floor(nLeft)+0.5,
Math.floor(nTop)+0.5,
nRoundedWidth,
nRoundedHeight);

ctx.strokeRect(
Math.ceil(nLeft)-0.5,
Math.ceil(nTop)-0.5,
nRoundedWidth,
nRoundedHeight);

});
});
}).call({\\"size\\":200,\\"value\\":\\"12345\\",\\"bgColor\\":\\"black\\",\\"fgColor\\":\\"white\\",\\"cells\\":[[true,true,true,true,true,true,true,false,false,false,true,false,true,false,true,true,true,true,true,true,true],[true,false,false,false,false,false,true,false,true,true,false,true,true,false,true,false,false,false,false,false,true],[true,false,true,true,true,false,true,false,false,true,false,true,true,false,true,false,true,true,true,false,true],[true,false,true,true,true,false,true,false,false,true,false,false,true,false,true,false,true,true,true,false,true],[true,false,true,true,true,false,true,false,false,false,false,false,true,false,true,false,true,true,true,false,true],[true,false,false,false,false,false,true,false,true,true,true,true,false,false,true,false,false,false,false,false,true],[true,true,true,true,true,true,true,false,true,false,true,false,true,false,true,true,true,true,true,true,true],[false,false,false,false,false,false,false,false,true,false,false,true,false,false,false,false,false,false,false,false,false],[false,false,false,false,true,true,true,true,false,false,false,false,false,false,true,true,false,false,false,true,false],[false,false,false,true,true,true,false,true,false,false,true,true,true,true,true,true,false,false,true,false,false],[true,false,false,true,true,true,true,true,true,true,false,false,true,true,false,true,true,true,true,false,false],[false,true,true,false,true,true,false,true,false,true,false,false,true,true,false,false,true,true,false,true,false],[false,false,true,false,false,false,true,true,true,false,false,true,true,true,false,false,false,true,false,true,true],[false,false,false,false,false,false,false,false,true,true,true,true,false,false,true,false,false,false,true,false,false],[true,true,true,true,true,true,true,false,true,true,false,true,true,false,false,false,false,false,true,false,false],[true,false,false,false,false,false,true,false,true,true,true,false,true,false,true,true,false,true,false,true,false],[true,false,true,true,true,false,true,false,true,true,true,true,true,true,false,true,true,true,false,false,true],[true,false,true,true,true,false,true,false,false,true,false,false,true,false,false,false,false,true,true,true,true],[true,false,true,true,true,false,true,false,false,false,true,true,true,false,false,true,true,true,false,false,false],[true,false,false,false,false,false,true,false,false,false,true,true,false,false,true,false,true,true,false,false,false],[true,true,true,true,true,true,true,false,false,true,false,false,false,true,false,true,false,true,false,false,true]]}, canvas);</script>",
            }
          }
          style={
            Object {
              "height": 200,
              "width": 200,
            }
          }
          underlayColor="transparent"
        />
      </View>
    </View>
  </View>
</View>
`;

exports[`send screen renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "justifyContent": "flex-start",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "height": 40,
          "marginHorizontal": 10,
          "marginVertical": 10,
        },
        Object {
          "marginTop": 30,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "flex-end",
          "height": 40,
          "marginRight": 10,
          "paddingTop": 2,
          "width": 80,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        To
      </Text>
    </View>
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      autoCorrect={false}
      multiline={true}
      onChangeText={[Function]}
      style={
        Array [
          Object {
            "backgroundColor": "white",
            "borderColor": "gray",
            "borderRadius": 6,
            "borderWidth": 1,
            "flex": 1,
            "fontSize": 13,
            "height": 26,
            "padding": 4,
          },
          Object {
            "height": 40,
          },
        ]
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "height": 40,
          "marginHorizontal": 10,
          "marginVertical": 10,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "flex-end",
          "height": 40,
          "marginRight": 10,
          "paddingTop": 2,
          "width": 80,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        Value (ETH)
      </Text>
    </View>
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      autoCorrect={false}
      keyboardType="decimal-pad"
      onChangeText={[Function]}
      style={
        Object {
          "backgroundColor": "white",
          "borderColor": "gray",
          "borderRadius": 6,
          "borderWidth": 1,
          "flex": 1,
          "fontSize": 13,
          "height": 26,
          "padding": 4,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <View
    style={
      Array [
        Object {
          "margin": 15,
        },
        undefined,
      ]
    }
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits="button"
      accessible={true}
      collapsable={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "flexDirection": "row",
          "height": 50,
          "opacity": 1,
          "overflow": "hidden",
          "width": 150,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <BVLinearGradient
        collapsable={false}
        colors={
          Array [
            4281228970,
            4281228970,
          ]
        }
        end={
          Object {
            "x": 1,
            "y": 1,
          }
        }
        locations={null}
        start={
          Object {
            "x": 0,
            "y": 0,
          }
        }
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "flex": 1,
            },
            undefined,
            Object {
              "borderRadius": 5,
            },
            null,
          ]
        }
      >
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "bottom": 0,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
                null,
              ]
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              collapsable={undefined}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "white",
                  "flexDirection": "row",
                  "fontSize": 14,
                  "letterSpacing": 2,
                  "opacity": 1,
                  "transform": Array [
                    Object {
                      "translateY": 0,
                    },
                  ],
                }
              }
            >
              Send
            </Text>
          </View>
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "bottom": 0,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
                Object {
                  "flexDirection": "column",
                },
              ]
            }
          >
            <View
              collapsable={undefined}
              style={
                Object {
                  "opacity": 0,
                  "transform": Array [
                    Object {
                      "translateY": -100,
                    },
                  ],
                }
              }
            >
              <RNSpinkit
                color={4294967295}
                size={25}
                style={
                  Array [
                    Object {
                      "height": 25,
                      "width": 25,
                    },
                    Array [
                      Object {},
                      undefined,
                    ],
                  ]
                }
                type="Wave"
              />
            </View>
            <Text
              accessible={true}
              allowFontScaling={true}
              collapsable={undefined}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "white",
                  "flexDirection": "row",
                  "fontSize": 14,
                  "letterSpacing": 2,
                  "opacity": 0,
                  "transform": Array [
                    Object {
                      "translateY": -50,
                    },
                  ],
                }
              }
            >
              Sending...
            </Text>
          </View>
        </View>
      </BVLinearGradient>
    </View>
  </View>
</View>
`;

exports[`whole app renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="#999999"
    hidesWhenStopped={true}
    size="large"
  />
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    Loading...
  </Text>
</View>
`;
